src/ucp/core/ucp_request.c

```diff
@@ -652,3 +652,31 @@ void ucp_request_purge_enqueue_cb(uct_pending_req_t *self, void *arg)
                   req->send.ep, req);
     ucs_queue_push(queue, (ucs_queue_elem_t*)&req->send.uct.priv);
 }
+
+ucs_status_t ucp_request_progress_wrapper(uct_pending_req_t *self)
+{
+    ucp_request_t *req       = ucs_container_of(self, ucp_request_t, send.uct);
+    const ucp_proto_t *proto = req->send.proto_config->proto;
+    uct_pending_callback_t progress_cb;
+    ucs_status_t status;
+
+    progress_cb = proto->progress[req->send.proto_stage];
+    ucp_trace_req(req,
+                  "progress %s {%ss} ep_cfg[%d] rkey_cfg[%d] offset %zu/%zu",
+                  proto->name, ucs_debug_get_symbol_name(progress_cb),
+                  req->send.proto_config->ep_cfg_index,
+                  req->send.proto_config->rkey_cfg_index,
+                  req->send.state.dt_iter.offset,
+                  req->send.state.dt_iter.length);
+
+    ucs_log_indent(1);
+    status = progress_cb(self);
+    if (status == UCS_OK) {
+        ucp_trace_req(req, "progress protocol %s returned OK", proto->name);
+    } else {
+        ucp_trace_req(req, "progress protocol %s returned: %s lane %d",
+                      proto->name, ucs_status_string(status), req->send.lane);
```

Comment:
can lane be a garbage in case of some error different than `NO_RESOURCE`?

Reply:
no since according to UCT pending API, callback may return any error to keep the request in pending https://github.com/openucx/ucx/blob/224f2ffdc0c0e72cf8ded0431addb83972684ba5/src/uct/api/uct_def.h#L528